File: src/cli/terminal.py
"""Terminal-integration helpers: tmux pane/window labels, kitty tab title,
and the start-of-run banner that prints version/runtime/cname to stderr.

All functions are best-effort: they no-op silently when the host
terminal feature isn't present (no TMUX, not a kitty session, no isatty,
or YOLO_NO_TMUX=1).
"""

import os
import platform
import subprocess
import sys
from pathlib import Path


def _get_project_name() -> str:
    """Return the jail project label: SM_PROJECT if set, else cwd basename."""
    return os.environ.get("SM_PROJECT") or Path.cwd().name


def _tmux_rename_window(name: str):
    """Rename the current tmux window. No-op if not in tmux or YOLO_NO_TMUX=1."""
    if os.environ.get("YOLO_NO_TMUX") == "1":
        return
    if os.environ.get("TMUX"):
        try:
            subprocess.run(
                ["tmux", "rename-window", name],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
        except Exception:
            pass


def _kitty_setup_jail_tab():
    """Set kitty tab title and color for jail indicator. Returns cleanup function or None."""
    if not os.environ.get("KITTY_PID") or not sys.stdin.isatty():
        return None

    project = _get_project_name()
    window_id = os.environ.get("KITTY_WINDOW_ID", "")
    match_arg = f"id:{window_id}" if window_id else "recent:0"

    def _kitten_run(cmd_args):
        try:
            subprocess.run(
                ["kitten", "@", *cmd_args],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
        except Exception:
            pass

    try:
        old_title = (
            subprocess.check_output(
                ["kitten", "@", "get-tab-title", "--match", match_arg],
                stderr=subprocess.DEVNULL,
            )
            .decode()
            .strip()
        )
    except Exception:
        old_title = ""

    try:
        subprocess.run(
            [
                "kitten",
                "@",
                "set-tab-title",
                "--match",
                match_arg,
                f"🔒 JAIL {project}",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
    except Exception:
        return None

    # Turn the tab red
    _kitten_run(
        [
            "set-tab-color",
            "--match",
            match_arg,
            "active_bg=#cc0000",
            "active_fg=#ffffff",
            "inactive_bg=#880000",
            "inactive_fg=#cccccc",
        ]
    )

    def restore():
        _kitten_run(["set-tab-title", "--match", match_arg, old_title or "bash"])
        # Reset tab colors to kitty.conf defaults
        _kitten_run(
            [
                "set-tab-color",
                "--match",
                match_arg,
                "active_bg=none",
                "active_fg=none",
                "inactive_bg=none",
                "inactive_fg=none",
            ]
        )

    return restore


def _tmux_setup_jail_pane():
    """Set tmux pane border indicators for the jail. Returns cleanup function."""
    if os.environ.get("YOLO_NO_TMUX") == "1":
        return None
    if not os.environ.get("TMUX") or not sys.stdin.isatty():
        return None

    pane = os.environ.get("TMUX_PANE", "")
    jail_dir = _get_project_name()

    def _tmux_opt(opt):
        try:
            r = subprocess.run(
                ["tmux", "show-option", "-pt", pane, opt],
                capture_output=True,
                text=True,
            )
            if r.returncode == 0 and r.stdout.strip():
                # Output is "option-name value" — extract value after first space
                parts = r.stdout.strip().split(None, 1)
                return parts[1] if len(parts) > 1 else ""
            return None
        except Exception:
            return None

    def _tmux_set(opt, val):
        try:
            subprocess.run(
                ["tmux", "set-option", "-pt", pane, opt, val],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
        except Exception:
            pass

    # Save old state
    old = {
        opt: _tmux_opt(opt)
        for opt in [
            "pane-border-style",
            "pane-active-border-style",
            "pane-border-status",
            "pane-border-format",
        ]
    }
    old_window = None
    old_auto_rename = None
    try:
        r = subprocess.run(
            ["tmux", "display-message", "-p", "#{window_name}"],
            capture_output=True,
            text=True,
        )
        old_window = r.stdout.strip() if r.returncode == 0 else None
        r = subprocess.run(
            ["tmux", "show-window-option", "-v", "automatic-rename"],
            capture_output=True,
            text=True,
        )
        old_auto_rename = r.stdout.strip() if r.returncode == 0 else None
    except Exception:
        pass

    # Set jail indicators
    _tmux_set("pane-border-style", "fg=red,bold")
    _tmux_set("pane-active-border-style", "fg=red,bold")
    _tmux_set("pane-border-status", "bottom")
    _tmux_set("pane-border-format", f" 🔒 JAIL {jail_dir} ")
    try:
        subprocess.run(
            ["tmux", "set-window-option", "automatic-rename", "off"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        subprocess.run(
            ["tmux", "rename-window", "JAIL"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
    except Exception:
        pass

    def restore():
        # Batch all tmux restores into a single command to minimize shutdown delay
        cmds = []
        for opt, val in old.items():
            if val is not None:
                cmds.append(f"set-option -pt {pane} {opt} {val}")
            else:
                cmds.append(f"set-option -put {pane} {opt}")
        if old_window:
            cmds.append(f"rename-window {old_window}")
        if old_auto_rename == "on":
            cmds.append("set-window-option automatic-rename on")
        if cmds:
            try:
                # Execute all restores in one tmux invocation using \;
                full_cmd = ["tmux"]
                for i, cmd in enumerate(cmds):
                    if i > 0:
                        full_cmd.append(";")
                    full_cmd.extend(cmd.split())
                subprocess.run(
                    full_cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
                )
            except Exception:
                pass

    return restore


def _print_startup_banner(
    version: str,
    runtime: str,
    cname: str,
    res_parts: "list[str] | None" = None,
    jail_version: "str | None" = None,
):
    """Print startup info to stderr for debugging and log sharing.

    ``version`` is the host CLI's version (what's running right now).
    ``jail_version``, if given, is the ``YOLO_VERSION`` baked into the
    already-running container — shown only when it differs from the
    host version, since that's the gap that silently causes stale
    shims / stale mounts / stale entrypoint logic on attach.
    """
    host_platform = f"{sys.platform}/{platform.machine()}"
    if jail_version and jail_version != version:
        ver_part = f"yolo-jail {version} (attached to jail built at {jail_version})"
    else:
        ver_part = f"yolo-jail {version}"
    parts = [ver_part, host_platform, runtime, cname]
    print(" | ".join(parts), file=sys.stderr)
    if res_parts:
        print(f"Resource limits: {', '.join(res_parts)}", file=sys.stderr)

File: internal/frontdoor/terminal.go
// Package frontdoor renders the terminal-facing half of the `yolo` run: the
// tmux/kitty jail indicators (SetupJailIndicator) and the start-of-run banner
// (StartupBanner). The banner's platform naming is x86_64/aarch64
// (platform.machine()), NOT Go's amd64/arm64.
package frontdoor

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
)

// projectName returns the jail label: $SM_PROJECT or the cwd basename.
func projectName() string {
	if p := os.Getenv("SM_PROJECT"); p != "" {
		return p
	}
	cwd, err := os.Getwd()
	if err != nil {
		return ""
	}
	return filepath.Base(cwd)
}

// SetupJailIndicator sets the terminal jail indicator (kitty tab or tmux pane
// border) and returns a restore func (or nil). Detection priority mirrors
// main(): KITTY_PID (and not TMUX) -> kitty; else tmux. YOLO_NO_TMUX=1 skips
// tmux. Only call this when NOT delegating (see the package doc).
func SetupJailIndicator() func() {
	if os.Getenv("KITTY_PID") != "" && os.Getenv("TMUX") == "" {
		return kittySetupJailTab()
	}
	return tmuxSetupJailPane()
}

func isattyStdin() bool {
	fi, err := os.Stdin.Stat()
	if err != nil {
		return false
	}
	return (fi.Mode() & os.ModeCharDevice) != 0
}

func kittenRun(args ...string) {
	cmd := exec.Command("kitten", append([]string{"@"}, args...)...)
	cmd.Stdout = nil
	cmd.Stderr = nil
	_ = cmd.Run()
}

func kittySetupJailTab() func() {
	if os.Getenv("KITTY_PID") == "" || !isattyStdin() {
		return nil
	}
	project := projectName()
	windowID := os.Getenv("KITTY_WINDOW_ID")
	matchArg := "recent:0"
	if windowID != "" {
		matchArg = "id:" + windowID
	}

	oldTitle := ""
	if out, err := exec.Command("kitten", "@", "get-tab-title", "--match", matchArg).Output(); err == nil {
		oldTitle = strings.TrimSpace(string(out))
	}

	set := exec.Command("kitten", "@", "set-tab-title", "--match", matchArg, "🔒 JAIL "+project)
	set.Stdout, set.Stderr = nil, nil
	if err := set.Run(); err != nil {
		return nil
	}
	kittenRun("set-tab-color", "--match", matchArg,
		"active_bg=#cc0000", "active_fg=#ffffff", "inactive_bg=#880000", "inactive_fg=#cccccc")

	return func() {
		restoreTitle := oldTitle
		if restoreTitle == "" {
			restoreTitle = "bash"
		}
		kittenRun("set-tab-title", "--match", matchArg, restoreTitle)
		kittenRun("set-tab-color", "--match", matchArg,
			"active_bg=none", "active_fg=none", "inactive_bg=none", "inactive_fg=none")
	}
}

func tmuxSetupJailPane() func() {
	if os.Getenv("YOLO_NO_TMUX") == "1" {
		return nil
	}
	if os.Getenv("TMUX") == "" || !isattyStdin() {
		return nil
	}
	pane := os.Getenv("TMUX_PANE")
	jailDir := projectName()

	tmuxOpt := func(opt string) (string, bool) {
		out, err := exec.Command("tmux", "show-option", "-pt", pane, opt).Output()
		if err != nil {
			return "", false
		}
		s := strings.TrimSpace(string(out))
		if s == "" {
			return "", false
		}
		parts := strings.SplitN(s, " ", 2)
		if len(parts) > 1 {
			return parts[1], true
		}
		return "", true
	}
	tmuxSet := func(opt, val string) {
		c := exec.Command("tmux", "set-option", "-pt", pane, opt, val)
		c.Stdout, c.Stderr = nil, nil
		_ = c.Run()
	}

	borderOpts := []string{"pane-border-style", "pane-active-border-style", "pane-border-status", "pane-border-format"}
	old := map[string]*string{}
	for _, opt := range borderOpts {
		if v, ok := tmuxOpt(opt); ok {
			vv := v
			old[opt] = &vv
		} else {
			old[opt] = nil
		}
	}
	oldWindow := ""
	if out, err := exec.Command("tmux", "display-message", "-p", "#{window_name}").Output(); err == nil {
		oldWindow = strings.TrimSpace(string(out))
	}
	oldAutoRename := ""
	if out, err := exec.Command("tmux", "show-window-option", "-v", "automatic-rename").Output(); err == nil {
		oldAutoRename = strings.TrimSpace(string(out))
	}

	tmuxSet("pane-border-style", "fg=red,bold")
	tmuxSet("pane-active-border-style", "fg=red,bold")
	tmuxSet("pane-border-status", "bottom")
	tmuxSet("pane-border-format", " 🔒 JAIL "+jailDir+" ")
	runQuiet("tmux", "set-window-option", "automatic-rename", "off")
	runQuiet("tmux", "rename-window", "JAIL")

	return func() {
		var cmds []string
		for _, opt := range borderOpts {
			if v := old[opt]; v != nil {
				cmds = append(cmds, "set-option -pt "+pane+" "+opt+" "+*v)
			} else {
				cmds = append(cmds, "set-option -put "+pane+" "+opt)
			}
		}
		if oldWindow != "" {
			cmds = append(cmds, "rename-window "+oldWindow)
		}
		if oldAutoRename == "on" {
			cmds = append(cmds, "set-window-option automatic-rename on")
		}
		if len(cmds) == 0 {
			return
		}
		full := []string{}
		for i, cmd := range cmds {
			if i > 0 {
				full = append(full, ";")
			}
			full = append(full, strings.Fields(cmd)...)
		}
		runQuiet("tmux", full...)
	}
}

func runQuiet(name string, args ...string) {
	c := exec.Command(name, args...)
	c.Stdout, c.Stderr = nil, nil
	_ = c.Run()
}

// hostPlatform returns "<goos>/<machine>" matching Python's
// f"{sys.platform}/{platform.machine()}", using the running GOOS/GOARCH.
func hostPlatform() string {
	return runtime.GOOS + "/" + platformMachine(runtime.GOOS, runtime.GOARCH)
}

// platformMachine maps Go's GOARCH to Python's platform.machine() spelling for
// the given GOOS. It is a pure function of (goos, goarch) so every OS/arch combo
// is unit-testable, not just the one the tests happen to run on. The spelling is
// Python's, NOT Go's amd64/arm64: amd64→x86_64 everywhere; arm64→aarch64 ONLY on
// Linux — on macOS/Apple Silicon platform.machine() is "arm64" (audit 2026-07-18
// §C: the unconditional arm64→aarch64 map was wrong on macOS and a test locked
// the bug). Any other GOARCH passes through unchanged.
func platformMachine(goos, goarch string) string {
	switch goarch {
	case "amd64":
		return "x86_64"
	case "arm64":
		if goos != "darwin" {
			return "aarch64" // Linux uname; macOS keeps arm64
		}
		return "arm64"
	default:
		return goarch
	}
}

// StartupBanner formats the start-of-run banner line(s) exactly as
// _print_startup_banner writes them to stderr. jailVersion is shown only when
// it differs from version. resParts, if non-empty, adds the resource-limits line.
func StartupBanner(version, runtime_, cname string, resParts []string, jailVersion string) string {
	var verPart string
	if jailVersion != "" && jailVersion != version {
		verPart = fmt.Sprintf("yolo-jail %s (attached to jail built at %s)", version, jailVersion)
	} else {
		verPart = "yolo-jail " + version
	}
	parts := []string{verPart, hostPlatform(), runtime_, cname}
	line := strings.Join(parts, " | ")
	if len(resParts) > 0 {
		line += "\nResource limits: " + strings.Join(resParts, ", ")
	}
	return line
}
